src/segment-loader.js

```diff
@@ -287,28 +315,74 @@
 
       // stop processing if the request was aborted
       if (!request.response) {
+        this.abort_();
         return;
       }
 
-      // calculate the download bandwidth
-      this.roundTrip = request.roundTripTime;
-      this.bandwidth = request.bandwidth;
-      this.bytesReceived += request.bytesReceived || 0;
+      if (request === this.xhr_.segmentXhr) {
+        segmentXhrRequest = this.xhr_.segmentXhr;
+        // the segment request is no longer outstanding
+        this.xhr_.segmentXhr = null;
 
-      segment = segmentInfo.playlist.segments[segmentInfo.mediaIndex];
-      if (segment.key) {
-        segmentInfo.encryptedBytes = new Uint8Array(request.response);
-      } else {
-        segmentInfo.bytes = new Uint8Array(request.response);
+        // calculate the download bandwidth based on segment request
+        this.roundTrip = request.roundTripTime;
+        this.bandwidth = request.bandwidth;
+        this.bytesReceived += request.bytesReceived || 0;
+
+        if (segment.key) {
+          segmentInfo.encryptedBytes = new Uint8Array(request.response);
+        } else {
+          segmentInfo.bytes = new Uint8Array(request.response);
+        }
       }
 
-      this.processResponse_();
+      if (request === this.xhr_.keyXhr) {
+        keyXhrRequest = this.xhr_.segmentXhr;
+        // the key request is no longer outstanding
+        this.xhr_.keyXhr = null;
+
+        view = new DataView(request.response);
+        segment.key.bytes = new Uint32Array([
+          view.getUint32(0),
+          view.getUint32(4),
+          view.getUint32(8),
+          view.getUint32(12)
+        ]);
+      }
+
+      if (!this.xhr_.segmentXhr && !this.xhr_.keyXhr) {
+        this.xhr_ = null;
+        this.processResponse_();
+      }
     },
 
     processResponse_: function() {
+      var segmentInfo, playlist, segment, mediaIndex, segIv;
+
       this.state = 'DECRYPTING';
 
-      this.handleSegment_();
+      segmentInfo = this.pendingSegment_;
+      playlist = segmentInfo.playlist;
+      mediaIndex = segmentInfo.mediaIndex;
+      segment = playlist.segments[mediaIndex];
+
+      if (segment.key) {
+        // this is an encrypted segment
+        // if the media sequence is greater than 2^32, the IV will be incorrect
+        // assuming 10s segments, that would be about 1300 years
+        segIv = segment.key.iv || new Uint32Array([0, 0, 0, mediaIndex + playlist.mediaSequence]);
+
+        // incrementally decrypt the segment
+        new videojs.Hls.Decrypter(segmentInfo.encryptedBytes,
+                                  segment.key.bytes,
+                                  segIv,
+                                  (function(err, bytes) {
```

Comment:
In the Decrypter code I only saw one use of done, and err is always set to null.